File: src/co/aospa/settings/gestures/TapScreenGestureInsidePreferenceController.java
/*
 * SPDX-FileCopyrightText: 2023 Yet Another AOSP Project
 * SPDX-License-Identifier: Apache-2.0
 */

package co.aospa.settings.gestures;

import android.content.Context;
import android.os.SystemProperties;
import android.provider.Settings;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;

import androidx.preference.Preference;
import androidx.preference.PreferenceScreen;

import co.aospa.framework.preference.SecureSettingSwitchPreference;

import com.android.settings.R;
import com.android.settings.core.PreferenceControllerMixin;
import com.android.settingslib.core.AbstractPreferenceController;
import com.android.settingslib.widget.MainSwitchPreference;

public class TapScreenGestureInsidePreferenceController extends AbstractPreferenceController
        implements PreferenceControllerMixin, OnCheckedChangeListener {

    private static final String KEY = "gesture_tap";
    private static final String AMBIENT_KEY = "doze_tap_gesture_ambient";

    private final Context mContext;
    private MainSwitchPreference mSwitch;
    private SecureSettingSwitchPreference mAmbientPref;

    public TapScreenGestureInsidePreferenceController(Context context) {
        super(context);
        mContext = context;
    }

    @Override
    public String getPreferenceKey() {
        return KEY;
    }

    @Override
    public void displayPreference(PreferenceScreen screen) {
        super.displayPreference(screen);
        mAmbientPref = screen.findPreference(AMBIENT_KEY);
        mSwitch = screen.findPreference(getPreferenceKey());
        mSwitch.setOnPreferenceClickListener(preference -> {
            final boolean enabled = Settings.Secure.getInt(mContext.getContentResolver(),
                    Settings.Secure.DOZE_TAP_SCREEN_GESTURE, 1) == 1;
            Settings.Secure.putInt(mContext.getContentResolver(),
                    Settings.Secure.DOZE_TAP_SCREEN_GESTURE,
                    enabled ? 0 : 1);
            updateAmbientEnablement(!enabled);
            return true;
        });
        mSwitch.addOnSwitchChangeListener(this);
        updateState(mSwitch);
    }

    public void setChecked(boolean isChecked) {
        if (mSwitch != null) {
            mSwitch.updateStatus(isChecked);
        }
        updateAmbientEnablement(isChecked);
    }

    @Override
    public void updateState(Preference preference) {
        final boolean enabled = Settings.Secure.getInt(mContext.getContentResolver(),
                Settings.Secure.DOZE_TAP_SCREEN_GESTURE, 1) == 1;
        setChecked(enabled);
    }

    @Override
    public boolean isAvailable() {
        return true;
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        Settings.Secure.putInt(mContext.getContentResolver(),
                Settings.Secure.DOZE_TAP_SCREEN_GESTURE, isChecked ? 1 : 0);
        SystemProperties.set("persist.sys.tap_gesture", isChecked ? "1" : "0");
        updateAmbientEnablement(isChecked);
    }

    private void updateAmbientEnablement(boolean enabled) {
        if (mAmbientPref == null) return;
        mAmbientPref.setEnabled(enabled);
    }
}

File: src/co/aospa/settings/gestures/TapScreenGestureInsidePreferenceController.kt
/*
 * SPDX-FileCopyrightText: 2023 Yet Another AOSP Project
 * SPDX-FileCopyrightText: 2025 Paranoid Android
 * SPDX-License-Identifier: Apache-2.0
 */

package co.aospa.settings.gestures

import android.content.Context
import android.os.SystemProperties
import android.provider.Settings
import android.widget.CompoundButton
import androidx.preference.Preference
import androidx.preference.PreferenceScreen
import co.aospa.framework.preference.SecureSettingSwitchPreference
import com.android.settings.R
import com.android.settingslib.core.AbstractPreferenceController
import com.android.settingslib.widget.MainSwitchPreference

class TapScreenGestureInsidePreferenceController(
    context: Context
) : AbstractPreferenceController(context), CompoundButton.OnCheckedChangeListener {

    private var mainPreference: MainSwitchPreference? = null
    private var ambientPreference: SecureSettingSwitchPreference? = null

    override fun getPreferenceKey(): String = KEY

    override fun displayPreference(screen: PreferenceScreen) {
        super.displayPreference(screen)
        mainPreference = screen.findPreference(getPreferenceKey())
        mainPreference?.addOnSwitchChangeListener(this)
        mainPreference?.let{updateState(it as Preference)}
        ambientPreference = screen.findPreference(AMBIENT_KEY)
    }

    fun setChecked(isChecked: Boolean) {
        mainPreference?.updateStatus(isChecked)
        updateAmbientEnablement(isChecked)
    }

    override fun updateState(preference: Preference) {
        val enabled = Settings.Secure.getInt(mContext.contentResolver, Settings.Secure.DOZE_TAP_SCREEN_GESTURE, 1) == 1
        setChecked(enabled)
    }

    override fun isAvailable(): Boolean = true

    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
        Settings.Secure.putInt(mContext.contentResolver, Settings.Secure.DOZE_TAP_SCREEN_GESTURE, if (isChecked) 1 else 0)
        SystemProperties.set("persist.sys.tap_gesture", if (isChecked) "1" else "0")
        updateAmbientEnablement(isChecked)
    }

    private fun updateAmbientEnablement(enabled: Boolean) {
        ambientPreference?.isEnabled = enabled
    }

    companion object {
        private const val KEY = "gesture_tap"
        private const val AMBIENT_KEY = "doze_tap_gesture_ambient"
    }
}
